src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs

```diff
@@ -893,7 +912,45 @@ private ZipArchiveEntry CreateEntry(ZipArchive package, string entryName, Compre
         private ZipArchiveEntry CreatePackageFileEntry(ZipArchive package, string entryName, DateTimeOffset timeOffset, CompressionLevel compressionLevel)
         {
             var entry = package.CreateEntry(entryName, compressionLevel);
-            entry.LastWriteTime = timeOffset;
+
+            // Please note: ZipArchive stream reader sometime changes LastWriteTime by another 1 second off than what "entry.LastWriteTime" has.
+            // The FAT filesystem of DOS has a timestamp resolution of only two seconds; ZIP file records mimic this.
+            // As a result, the built -in timestamp resolution of files in a ZIP archive is only two seconds, though extra fields can be used to store more precise timestamps.
+            if (timeOffset < ZipFormatMinDate)
+            {
+                if (!_zipFormatCorrected)
+                {
+                    _zipFormatCorrected = true;
+
+                    _logger.Log(PackagingLogMessage.CreateMessage(
+                                    string.Format(
+                                        CultureInfo.CurrentCulture,
+                                        Strings.ZipFileTimeStampModified, entryName, timeOffset.DateTime.ToShortDateString(), ZipFormatMinDate.ToShortDateString(), ZipFormatMinDate.ToShortDateString(), ZipFormatMaxDate.ToShortDateString()),
+                                        LogLevel.Information));
+                }
+
+                entry.LastWriteTime = ZipFormatMinDate;
+            }
+            else if (timeOffset > ZipFormatMaxDate)
+            {
+                if (!_zipFormatCorrected)
```

Comment:
Why log only once?
We should log for *all* files that were adjusted.

Reply:
Because of this [comment](https://github.com/NuGet/NuGet.Client/pull/3793#discussion_r544603589) to limit spamming.

Reply:
The method `WriteFiles` iterates the files in the package and calls `CreatePart` for each. If you create a collection in this method, containing all the filenames with "out of range" dates in this method, then you can create a single message with all the affected files. As an example, see [this pack validation rule](https://github.com/NuGet/NuGet.Client/blob/4991066d8d78053b389b07819afba78b88e66340/src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs#L41-L56)

Reply:
I agree with @zivkan 

Limit spamming means omitting repetitive information.

Reply:
Other warnings had a block of text, followed by a list of shorter messages for each individual instance. So, you could think about changing the warning to something like:

```
The zip format supports a limited date range. The following files are outside the supported range:
'file1' changed from '1979-12-31' to '1980-01-01'
'file2' changed from '2108-01-01' to '2107-12-31'
```

This way we can avoid repeating the "out of range for what the zip format supports" text.

Also, if you were to do this in Visual Studio, you should see only a single entry in the error list, not one per file. I'm not sure if it reduces a list of warnings in a binlong.

Having said this, I was thinking about it over the weekend, and I'm feeling much less strongly about this. Consider a c# syntax error instead. If I have the same syntax error in multiple places in a single file, Roslyn is going to raise one warning per instance, not one warning per file or one warning per project. One difference is that in Visual Studio you can click the warning and go to the line of code, whereas NuGet doesn't. Even if NuGet had the capability to determine a filename and line number, for "go to error" in general, I'm not sure it makes sense for this specific warning. It probably makes sense here to have a single message with a list of affected files.

As mentioned in another comment, we should have an NU warning code, and a docs issue to document the NU code.